File: expressjs/src/engine/nested-write.ts
import type { Queryable, Store } from "../store/postgres.js";
import { queryRow, queryRows, exec } from "../store/postgres.js";
import type { Entity } from "../metadata/types.js";
import type { Registry } from "../metadata/registry.js";
import { getField, fieldNames } from "../metadata/types.js";
import type { ErrorDetail } from "./errors.js";
import { AppError, validationError } from "./errors.js";
import type { RelationWrite } from "./writer.js";
import {
  buildInsertSQL,
  buildUpdateSQL,
  validateFields,
  separateFieldsAndRelations,
} from "./writer.js";
import { executeChildWrite } from "./diff.js";
import { evaluateRules } from "./rules.js";
import { evaluateStateMachines } from "./state-machine.js";
import { triggerWorkflows } from "./workflow.js";
import { fireAsyncWebhooks, fireSyncWebhooks } from "./webhook.js";
import { getInstrumenter } from "../instrument/instrument.js";

export interface WritePlan {
  isCreate: boolean;
  entity: Entity;
  fields: Record<string, any>;
  id: any;
  childOps: RelationWrite[];
  user: { id: string; roles: string[] } | null;
}

export function planWrite(
  entity: Entity,
  registry: Registry,
  body: Record<string, any>,
  existingID: any | null,
): WritePlan {
  const [fields, relWrites, unknownKeys] = separateFieldsAndRelations(
    entity,
    registry,
    body,
  );

  // Reject unknown keys
  if (unknownKeys.length > 0) {
    const errs: ErrorDetail[] = unknownKeys.map((key) => ({
      field: key,
      rule: "unknown",
      message: `Unknown field or relation: ${key}`,
    }));
    throw validationError(errs);
  }

  const isCreate = existingID == null;

  const validationErrs = validateFields(entity, fields, isCreate);
  if (validationErrs.length > 0) {
    throw validationError(validationErrs);
  }

  const plan: WritePlan = {
    isCreate,
    entity,
    fields,
    id: existingID,
    childOps: Array.from(relWrites.values()),
    user: null,
  };

  return plan;
}

export async function executeWritePlan(
  store: Store,
  registry: Registry,
  plan: WritePlan,
): Promise<Record<string, any>> {
  const span = getInstrumenter().startSpan("engine", "writer", "nested_write.execute");
  span.setEntity(plan.entity.name, plan.id ?? undefined);
  span.setMetadata("operation", plan.isCreate ? "create" : "update");
  const client = await store.beginTx();
  try {
    // Evaluate rules (field → expression → computed)
    let old: Record<string, any> = {};
    if (!plan.isCreate) {
      try {
        old = await fetchRecord(client, plan.entity, plan.id);
      } catch {
        // ignore if old record not found
      }
    }

    const ruleErrs = evaluateRules(
      registry,
      plan.entity.name,
      "before_write",
      plan.fields,
      old,
      plan.isCreate,
    );
    if (ruleErrs.length > 0) {
      throw validationError(ruleErrs);
    }

    // Evaluate state machines (after rules, before SQL write)
    const smErrs = evaluateStateMachines(
      registry,
      plan.entity.name,
      plan.fields,
      old,
      plan.isCreate,
    );
    if (smErrs.length > 0) {
      throw validationError(smErrs);
    }

    // Resolve file fields: UUID string → JSONB metadata object
    await resolveFileFields(client, plan.entity, plan.fields);

    let parentID: any;

    if (plan.isCreate) {
      const [sql, params] = buildInsertSQL(plan.entity, plan.fields);
      const row = await queryRow(client, sql, params);
      parentID = row[plan.entity.primary_key.field];
    } else {
      parentID = plan.id;
      const [sql, params] = buildUpdateSQL(plan.entity, plan.id, plan.fields);
      if (sql) {
        await exec(client, sql, params!);
      }
    }

    for (const childOp of plan.childOps) {
      await executeChildWrite(client, registry, parentID, childOp);
    }

    // Pre-commit: fire sync (before_write) webhooks
    const action = plan.isCreate ? "create" : "update";
    await fireSyncWebhooks(client, registry, "before_write", plan.entity.name, action, plan.fields, old, plan.user);

    await client.query("COMMIT");

    // Fetch the full record
    const result = await fetchRecord(store.pool, plan.entity, parentID);

    // Post-commit: trigger workflows for state transitions
    const machines = registry.getStateMachinesForEntity(plan.entity.name);
    for (const sm of machines) {
      const oldState = old[sm.field] != null ? String(old[sm.field]) : "";
      const newState = result[sm.field] != null ? String(result[sm.field]) : "";
      if (oldState !== newState && newState !== "") {
        triggerWorkflows(store, registry, plan.entity.name, sm.field, newState, result, parentID).catch((err) => {
          console.error(`ERROR: triggering workflows for ${plan.entity.name}.${sm.field} -> ${newState}:`, err);
        });
      }
    }

    // Post-commit: fire async (after_write) webhooks
    fireAsyncWebhooks(store, registry, "after_write", plan.entity.name, action, result, old, plan.user);

    span.setStatus("ok");
    return result;
  } catch (err) {
    await client.query("ROLLBACK");
    span.setStatus("error");
    span.setMetadata("error", (err as Error).message);
    throw err;
  } finally {
    client.release();
    span.end();
  }
}

export async function fetchRecord(
  q: Queryable,
  entity: Entity,
  id: any,
): Promise<Record<string, any>> {
  const columns = fieldNames(entity);
  if (entity.soft_delete && !getField(entity, "deleted_at")) {
    columns.push("deleted_at");
  }

  let sql = `SELECT ${columns.join(", ")} FROM ${entity.table} WHERE ${entity.primary_key.field} = $1`;
  if (entity.soft_delete) {
    sql += " AND deleted_at IS NULL";
  }

  return queryRow(q, sql, [id]);
}

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/**
 * For each file-type field whose value is a UUID string, resolve it to
 * the full JSONB metadata object {id, filename, size, mime_type, url}.
 * If the value is already an object, pass through unchanged.
 */
async function resolveFileFields(
  q: Queryable,
  entity: Entity,
  fields: Record<string, any>,
): Promise<void> {
  for (const f of entity.fields) {
    if (f.type !== "file") continue;
    const val = fields[f.name];
    if (val === undefined || val === null) continue;
    // If already an object (full metadata), pass through
    if (typeof val === "object") continue;
    // If it's a UUID string, resolve from _files
    const strVal = String(val);
    if (!UUID_RE.test(strVal)) continue;

    const rows = await queryRows(
      q,
      "SELECT id, filename, size, mime_type FROM _files WHERE id = $1",
      [strVal],
    );
    if (rows.length === 0) {
      throw new AppError("NOT_FOUND", 404, `File ${strVal} not found`);
    }
    const fileRow = rows[0];
    fields[f.name] = {
      id: fileRow.id,
      filename: fileRow.filename,
      size: Number(fileRow.size),
      mime_type: fileRow.mime_type,
    };
  }
}

File: golang/internal/engine/nested_write.go
package engine

import (
	"context"
	"fmt"
	"regexp"
	"strings"
	"unicode"

	"rocket-backend/internal/instrument"
	"rocket-backend/internal/metadata"
	"rocket-backend/internal/store"

	"golang.org/x/text/unicode/norm"
)

var uuidRE = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

// WritePlan describes the full set of operations for a write request.
type WritePlan struct {
	IsCreate  bool
	Entity    *metadata.Entity
	Fields    map[string]any
	ID        any // nil for create, set for update
	ChildOps  []*RelationWrite
	User      *metadata.UserContext
}

// PlanWrite builds a WritePlan from the request body without executing any SQL.
func PlanWrite(entity *metadata.Entity, reg *metadata.Registry, body map[string]any, existingID any) (*WritePlan, []ErrorDetail) {
	fields, relWrites, unknownKeys := SeparateFieldsAndRelations(entity, reg, body)

	// Reject unknown keys
	if len(unknownKeys) > 0 {
		var errs []ErrorDetail
		for _, key := range unknownKeys {
			errs = append(errs, ErrorDetail{
				Field:   key,
				Rule:    "unknown",
				Message: fmt.Sprintf("Unknown field or relation: %s", key),
			})
		}
		return nil, errs
	}

	isCreate := existingID == nil

	// Validate fields
	validationErrs := ValidateFields(entity, fields, isCreate)
	if len(validationErrs) > 0 {
		return nil, validationErrs
	}

	plan := &WritePlan{
		IsCreate: isCreate,
		Entity:   entity,
		Fields:   fields,
		ID:       existingID,
	}

	for _, rw := range relWrites {
		plan.ChildOps = append(plan.ChildOps, rw)
	}

	return plan, nil
}

// ExecuteWritePlan runs the planned operations inside a single transaction.
// Returns the created/updated record.
func ExecuteWritePlan(ctx context.Context, s *store.Store, reg *metadata.Registry, plan *WritePlan) (map[string]any, error) {
	ctx, span := instrument.GetInstrumenter(ctx).StartSpan(ctx, "engine", "writer", "nested_write.execute")
	defer span.End()
	span.SetEntity(plan.Entity.Name, fmt.Sprintf("%v", plan.ID))

	tx, err := s.BeginTx(ctx)
	if err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return nil, fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback() //nolint:errcheck

	// Evaluate rules (field -> expression -> computed)
	var old map[string]any
	if !plan.IsCreate {
		old, _ = fetchRecord(ctx, tx, plan.Entity, plan.ID, s.Dialect)
	}
	if old == nil {
		old = map[string]any{}
	}

	ruleErrs := EvaluateRules(ctx, reg, plan.Entity.Name, "before_write", plan.Fields, old, plan.IsCreate)
	if len(ruleErrs) > 0 {
		span.SetStatus("error")
		return nil, ValidationError(ruleErrs)
	}

	// Evaluate state machines (after rules, before SQL write)
	smErrs := EvaluateStateMachines(ctx, reg, plan.Entity.Name, plan.Fields, old, plan.IsCreate)
	if len(smErrs) > 0 {
		span.SetStatus("error")
		return nil, ValidationError(smErrs)
	}

	// Auto-generate slug if configured
	if err := autoGenerateSlug(ctx, tx, plan.Entity, s.Dialect, plan.Fields, plan.IsCreate, old, plan.ID); err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return nil, err
	}

	// Resolve file fields: UUID string -> JSONB metadata object
	if err := resolveFileFields(ctx, tx, plan.Entity, plan.Fields, s.Dialect); err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return nil, fmt.Errorf("resolve file fields: %w", err)
	}

	var parentID any

	if plan.IsCreate {
		// INSERT parent
		sql, params := BuildInsertSQL(plan.Entity, plan.Fields, s.Dialect)
		row, err := store.QueryRow(ctx, tx, sql, params...)
		if err != nil {
			span.SetStatus("error")
			span.SetMetadata("error", err.Error())
			return nil, fmt.Errorf("insert %s: %w", plan.Entity.Table, err)
		}
		parentID = row[plan.Entity.PrimaryKey.Field]
	} else {
		// UPDATE parent
		parentID = plan.ID
		sql, params := BuildUpdateSQL(plan.Entity, plan.ID, plan.Fields, s.Dialect)
		if sql != "" {
			if _, err := store.Exec(ctx, tx, sql, params...); err != nil {
				span.SetStatus("error")
				span.SetMetadata("error", err.Error())
				return nil, fmt.Errorf("update %s: %w", plan.Entity.Table, err)
			}
		}
	}

	// Execute child writes
	for _, childOp := range plan.ChildOps {
		if err := ExecuteChildWrite(ctx, tx, s.Dialect, reg, parentID, childOp); err != nil {
			span.SetStatus("error")
			span.SetMetadata("error", err.Error())
			return nil, fmt.Errorf("child write for %s: %w", childOp.Relation.Name, err)
		}
	}

	// Pre-commit: fire sync (before_write) webhooks
	action := "update"
	if plan.IsCreate {
		action = "create"
	}
	if err := FireSyncWebhooks(ctx, tx, s.Dialect, reg, "before_write", plan.Entity.Name, action, plan.Fields, old, plan.User); err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return nil, fmt.Errorf("sync webhook: %w", err)
	}

	// Commit
	if err := tx.Commit(); err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return nil, fmt.Errorf("commit: %w", err)
	}

	// Fetch the full record
	record, err := fetchRecord(ctx, s.DB, plan.Entity, parentID, s.Dialect)
	if err != nil {
		span.SetStatus("error")
		span.SetMetadata("error", err.Error())
		return nil, err
	}

	// Post-commit: trigger workflows for state transitions
	for _, sm := range reg.GetStateMachinesForEntity(plan.Entity.Name) {
		oldState := ""
		if v, ok := old[sm.Field]; ok && v != nil {
			oldState = fmt.Sprintf("%v", v)
		}
		newState := ""
		if v, ok := plan.Fields[sm.Field]; ok && v != nil {
			newState = fmt.Sprintf("%v", v)
		}
		if newState != "" && oldState != newState {
			TriggerWorkflows(ctx, s, reg, plan.Entity.Name, sm.Field, newState, record, parentID)
		}
	}

	// Post-commit: fire async (after_write) webhooks
	FireAsyncWebhooks(ctx, s, reg, "after_write", plan.Entity.Name, action, record, old, plan.User)

	span.SetStatus("ok")
	return record, nil
}

func fetchRecord(ctx context.Context, q store.Querier, entity *metadata.Entity, id any, dialect store.Dialect) (map[string]any, error) {
	columns := entity.FieldNames()
	if entity.SoftDelete && entity.GetField("deleted_at") == nil {
		columns = append(columns, "deleted_at")
	}

	softDeleteClause := ""
	if entity.SoftDelete {
		softDeleteClause = " AND deleted_at IS NULL"
	}

	// If entity has a slug config and the param doesn't look like the PK type, try slug first
	idStr := fmt.Sprintf("%v", id)
	if entity.Slug != nil && !looksLikePK(entity, idStr) {
		slugSQL := fmt.Sprintf("SELECT %s FROM %s WHERE %s = %s%s",
			joinColumns(columns), entity.Table, entity.Slug.Field, dialect.Placeholder(1), softDeleteClause)
		row, err := store.QueryRow(ctx, q, slugSQL, idStr)
		if err == nil {
			return row, nil
		}
		// slug lookup failed, fall through to PK lookup
	}

	sql := fmt.Sprintf("SELECT %s FROM %s WHERE %s = %s%s",
		joinColumns(columns), entity.Table, entity.PrimaryKey.Field, dialect.Placeholder(1), softDeleteClause)

	return store.QueryRow(ctx, q, sql, id)
}

var intRE = regexp.MustCompile(`^\d+$`)

func looksLikePK(entity *metadata.Entity, value string) bool {
	switch entity.PrimaryKey.Type {
	case "uuid":
		return uuidRE.MatchString(value)
	case "int", "integer", "bigint":
		return intRE.MatchString(value)
	default:
		return false // string PKs — can't distinguish, always try slug first
	}
}

// Slugify converts a string into a URL-friendly slug.
func Slugify(text string) string {
	// Normalize unicode and strip accents
	result := norm.NFD.String(text)
	var b strings.Builder
	for _, r := range result {
		if unicode.Is(unicode.Mn, r) {
			continue // skip combining marks (accents)
		}
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		} else if r >= 'A' && r <= 'Z' {
			b.WriteRune(r + 32) // lowercase
		} else {
			b.WriteByte('-')
		}
	}
	// Collapse multiple hyphens and trim
	s := b.String()
	for strings.Contains(s, "--") {
		s = strings.ReplaceAll(s, "--", "-")
	}
	s = strings.Trim(s, "-")
	return s
}

func generateUniqueSlug(ctx context.Context, q store.Querier, entity *metadata.Entity, dialect store.Dialect, baseSlug string, excludeID any) (string, error) {
	slugField := entity.Slug.Field
	softDeleteClause := ""
	if entity.SoftDelete {
		softDeleteClause = " AND deleted_at IS NULL"
	}

	checkSQL := fmt.Sprintf("SELECT 1 FROM %s WHERE %s = %s%s", entity.Table, slugField, dialect.Placeholder(1), softDeleteClause)
	excludeClause := ""
	if excludeID != nil {
		excludeClause = fmt.Sprintf(" AND %s != %s", entity.PrimaryKey.Field, dialect.Placeholder(2))
		checkSQL = fmt.Sprintf("SELECT 1 FROM %s WHERE %s = %s%s%s", entity.Table, slugField, dialect.Placeholder(1), softDeleteClause, excludeClause)
	}

	// Try base slug
	var params []any
	if excludeID != nil {
		params = []any{baseSlug, excludeID}
	} else {
		params = []any{baseSlug}
	}
	rows, err := store.QueryRows(ctx, q, checkSQL+" LIMIT 1", params...)
	if err != nil {
		return "", err
	}
	if len(rows) == 0 {
		return baseSlug, nil
	}

	// Append incrementing suffix
	for i := 2; i <= 100; i++ {
		candidate := fmt.Sprintf("%s-%d", baseSlug, i)
		params[0] = candidate
		rows, err = store.QueryRows(ctx, q, checkSQL+" LIMIT 1", params...)
		if err != nil {
			return "", err
		}
		if len(rows) == 0 {
			return candidate, nil
		}
	}

	// Fallback: should not normally reach here
	return fmt.Sprintf("%s-%d", baseSlug, 101), nil
}

func autoGenerateSlug(ctx context.Context, q store.Querier, entity *metadata.Entity, dialect store.Dialect, fields map[string]any, isCreate bool, old map[string]any, existingID any) error {
	slugCfg := entity.Slug
	if slugCfg == nil || slugCfg.Source == "" {
		return nil
	}

	// If slug is explicitly provided, skip auto-generation
	if val, ok := fields[slugCfg.Field]; ok && val != nil && fmt.Sprintf("%v", val) != "" {
		return nil
	}

	sourceVal, hasSource := fields[slugCfg.Source]
	if !hasSource || sourceVal == nil || fmt.Sprintf("%v", sourceVal) == "" {
		return nil
	}

	if isCreate {
		slug, err := generateUniqueSlug(ctx, q, entity, dialect, Slugify(fmt.Sprintf("%v", sourceVal)), nil)
		if err != nil {
			return fmt.Errorf("generate slug: %w", err)
		}
		fields[slugCfg.Field] = slug
	} else if slugCfg.RegenerateOnUpdate {
		// Only regenerate if source field changed
		oldSourceVal := fmt.Sprintf("%v", old[slugCfg.Source])
		newSourceVal := fmt.Sprintf("%v", sourceVal)
		if oldSourceVal == newSourceVal {
			return nil
		}
		slug, err := generateUniqueSlug(ctx, q, entity, dialect, Slugify(newSourceVal), existingID)
		if err != nil {
			return fmt.Errorf("generate slug: %w", err)
		}
		fields[slugCfg.Field] = slug
	}

	return nil
}

func joinColumns(cols []string) string {
	result := ""
	for i, c := range cols {
		if i > 0 {
			result += ", "
		}
		result += c
	}
	return result
}

// resolveFileFields converts UUID strings in file-type fields to full JSONB metadata objects.
// If the value is already a map (full metadata), it passes through unchanged.
func resolveFileFields(ctx context.Context, q store.Querier, entity *metadata.Entity, fields map[string]any, dialect store.Dialect) error {
	for _, f := range entity.Fields {
		if f.Type != "file" {
			continue
		}
		val, ok := fields[f.Name]
		if !ok || val == nil {
			continue
		}
		// If already a map (full metadata object), pass through
		if _, isMap := val.(map[string]any); isMap {
			continue
		}
		// If it's a UUID string, resolve from _files
		strVal := fmt.Sprintf("%v", val)
		if !uuidRE.MatchString(strVal) {
			continue
		}

		row, err := store.QueryRow(ctx, q,
			fmt.Sprintf("SELECT id, filename, size, mime_type FROM _files WHERE id = %s", dialect.Placeholder(1)), strVal)
		if err != nil {
			return NewAppError("NOT_FOUND", 404, fmt.Sprintf("File %s not found", strVal))
		}

		fields[f.Name] = map[string]any{
			"id":        row["id"],
			"filename":  row["filename"],
			"size":      row["size"],
			"mime_type": row["mime_type"],
		}
	}
	return nil
}
